Mods/SML/Source/SML/Private/Patching/BlueprintHookManager.cpp

```diff
@@ -35,44 +35,35 @@ void DebugDumpFunctionScriptCode(UFunction* Function, int32 HookOffset, const FS
 }
 #endif
 
-void UBlueprintHookManager::InstallBlueprintHook(UFunction* Function, int32 HookOffset) {
+void UBlueprintHookManager::InstallBlueprintHook(UFunction* Function, const int32 RequestedHookOffset, const int32 ResolvedHookOffset) {
 	TArray<uint8>& OriginalCode = Function->Script;
-	checkf(OriginalCode.Num() > HookOffset, TEXT("Invalid hook: HookOffset > Script.Num()"));
+	fgcheckf(OriginalCode.Num() > ResolvedHookOffset, TEXT("Invalid hook: Resolved HookOffset > Script.Num()"));
 
 #if DEBUG_BLUEPRINT_HOOKING
-	DebugDumpFunctionScriptCode(Function, HookOffset, TEXT("BeforeHook"));
+	DebugDumpFunctionScriptCode(Function, RequestedHookOffset, ResolvedHookOffset, TEXT("BeforeHook"));
 #endif
-	
-	//Minimum amount of bytes required to insert unconditional jump with code offset
-	const int32 MinBytesRequired = 1 + sizeof(CodeSkipSizeType);
 
+	// Will throw an error if the resolved hook offset is not at properly-aligned, parseable opcode
 	FSMLKismetBytecodeDisassembler Disassembler;
-	int32 BytesAvailable = 0;
-	
-	//Walk over statements until we collect enough bytes for a replacement
-	//(or until we consumed all statements in the function's code)
-	while (BytesAvailable < MinBytesRequired && (HookOffset + BytesAvailable) < OriginalCode.Num()) {
-		const int32 CurrentStatementIndex = HookOffset + BytesAvailable;
-		int32 OutStatementLength;
-		
-		const bool bValid = Disassembler.GetStatementLength(Function, CurrentStatementIndex, OutStatementLength);
-		checkf(bValid, TEXT("Provided hook offset is not a valid statement index: %d"), HookOffset);
-		BytesAvailable += OutStatementLength;
+	Disassembler.SerializeStatement(Function, ResolvedHookOffset);
+
+	// First we go over the existing code and add UBlueprintHookManager::JumpBytesRequired to all
+	// the offsets. Afterwards we will move the relevant code and add the jump.  Doing it in this
+	// order just means we don't have to worry about errantly changing the jump offset we will add.
+	TArray<TSharedPtr<FJsonValue>> DisassembledFunction = Disassembler.SerializeFunction(Function);
+	for (TSharedPtr<FJsonValue> JsonValue : DisassembledFunction) {
+		ModifyJumpTargetsForNewHookOffset(OriginalCode, JsonValue->AsObject(), ResolvedHookOffset);
 	}
 
-	//Check that we collected enough bytes
-	if (BytesAvailable < MinBytesRequired) {
-		//If we are here, it means we consumed all the statements in the function's code
-		//And still don't have enough space for inserting a jump. In that case, we append additional
-		//EX_EndOfScript instructions until we have enough place
-		const int32 BytesToAppend = MinBytesRequired - BytesAvailable;
-		OriginalCode.AddUninitialized(BytesToAppend);
-		FPlatformMemory::Memset(&OriginalCode[OriginalCode.Num() - BytesToAppend], EX_EndOfScript, BytesToAppend);
-		BytesAvailable = MinBytesRequired;
+	// Add enough room to add the jump
+	OriginalCode.AddUninitialized(UBlueprintHookManager::JumpBytesRequired);
```

Comment:
TArrays have an `InsertUninitialized` function you can use instead of moving instructions around, would it be possible to use it here?

Reply:
Oh, awesome.  I changed to that and it works perfectly.

Reply:
Ack